[web-animations] contain-intrinsic-* properties should support animations

cathiechen opened the pull request:
#### 41b432cafb957d1fdd83f96ffc767a15501de1ea
<pre>
[web-animations] contain-intrinsic-* properties should support animations
<a href="https://bugs.webkit.org/show_bug.cgi?id=241184">https://bugs.webkit.org/show_bug.cgi?id=241184</a>

Reviewed by Antoine Quint.

This patch support animations for contain-intrinsic-size related properties. As the value of contain-intrinsic-size is consitent of ContainIntrinsicSizeType
and an std::optional&lt;LayoutUnit&gt; value. In this patch, we added OptionalLengthPropertyWrapper to handle the optinal length, and ContainIntrinsiclLengthPropertyWrapper
to handle ContainIntrinsicSizeType.

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/animation/contain-intrinsic-size-interpolation-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::OptionalLengthPropertyWrapper::OptionalLengthPropertyWrapper):
(WebCore::OptionalLengthPropertyWrapper::canInterpolate const):
(WebCore::OptionalLengthPropertyWrapper::blend const):
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):

Canonical link: <a href="https://commits.webkit.org/256028@main">https://commits.webkit.org/256028@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/569dd29ebdfa7a377bad56f996df836fe7c6e571

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94351 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3529 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/27278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104018 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164292 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98349 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3576 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31735 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86688 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100018 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100021 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2564 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80746 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29611 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84486 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/27278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72502 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38130 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/27278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36007 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/27278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4167 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39894 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41844 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/27278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1079,6 +1079,79 @@ class LengthVariantPropertyWrapper final : public PropertyWrapperGetter<const T&
+        return OptionalLengthPropertyWrapper::canInterpolate(from, to, operation);
+    }
+
+    void blend(RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, const CSSPropertyBlendingContext& context) const
```
I don't really know if it matters given the class itself is marked `final`, but `canInterpolate` is marked `final` while this method isn't.

cathiechen replied:
Done, thanks! Yeah, I think it should be marked `final`(Or `override` if not).

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1079,6 +1079,79 @@ class LengthVariantPropertyWrapper final : public PropertyWrapperGetter<const T&
+    void blend(RenderStyle& destination, const RenderStyle& from, const RenderStyle& to, const CSSPropertyBlendingContext& context) const
+    {
+        if (context.isDiscrete) {
+            (destination.*m_setter)(context.progress < 0.5 ? this->value(from) : this->value(to));
```
You don't need to check for the `0.5` boundary here since you're guaranteed to get a `0` or `1` value since this is a discrete interpolation. So this can be simplified to:

```
(destination.*m_setter)(this->value(context.progress ? to : from))
```

You can also add this:

```
ASSERT(!context.progress || context.progress == 1);
```

cathiechen replied:
Done, thanks!

graouts reviewed (approved):
Do you know if the remaining animation failures are only related to the contested use of `inherit`?

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1079,6 +1079,80 @@ class LengthVariantPropertyWrapper final : public PropertyWrapperGetter<const T&
+    {
+        if (context.isDiscrete) {
+            ASSERT(!context.progress || context.progress == 1);
+            (destination.*m_setter)(context.progress ? this->value(from) : this->value(to));
```
You have `from` and `to` the wrong way around.

cathiechen replied:
Oops! Done!

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1079,6 +1079,80 @@ class LengthVariantPropertyWrapper final : public PropertyWrapperGetter<const T&
+protected:
+    bool canInterpolate(const RenderStyle& from, const RenderStyle& to, CompositeOperation) const override
+    {
+        if (!this->value(from).has_value() || !this->value(to).has_value())
```
So I believe you can just write:

```
if (!this->value(from) || !this->value(to))
```

cathiechen replied:
Done, thanks!

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1079,6 +1079,80 @@ class LengthVariantPropertyWrapper final : public PropertyWrapperGetter<const T&
+            return false;
+
+        bool isLengthPercentage = m_flags.contains(Flags::IsLengthPercentage);
+        return canInterpolateLengths(this->value(from).value(), this->value(to).value(), isLengthPercentage);
```
And here you can do:

```
return canInterpolateLengths(*this->value(from), *this->value(to), isLengthPercentage);
```

cathiechen replied:
The compiler complains about this change, looks like it can convert `std::optional<Length>` to `Length` implicitly. 
```
static bool canInterpolateLengths(const Length& from, const Length& to, bool isLengthPercentage)
```

cathiechen replied:
Ah, looks like I missed adding the operator *... It's done now.

graouts commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -1079,6 +1079,80 @@ class LengthVariantPropertyWrapper final : public PropertyWrapperGetter<const T&
+        }
+
+        auto valueRange = m_flags.contains(Flags::NegativeLengthsAreInvalid) ? ValueRange::NonNegative : ValueRange::All;
+        (destination.*m_setter)(blendFunc(this->value(from).value(), this->value(to).value(), context, valueRange));
```
Likewise, you should be able to do:

```
(destination.*m_setter)(blendFunc(*this->value(from), *this->value(to), context, valueRange));
```

cathiechen replied:
Ditto, the compiler complains about this change:)
```
static inline Length blendFunc(const Length& from, const Length& to, const CSSPropertyBlendingContext& context, ValueRange valueRange = ValueRange::All)

```

cathiechen replied:
done

cathiechen committed:
[web-animations] contain-intrinsic-* properties should support animations
https://bugs.webkit.org/show_bug.cgi?id=241184

Reviewed by Antoine Quint.

This patch support animations for contain-intrinsic-size related properties. As the value of contain-intrinsic-size is consitent of ContainIntrinsicSizeType
and an std::optional<LayoutUnit> value. In this patch, we added OptionalLengthPropertyWrapper to handle the optinal length, and ContainIntrinsiclLengthPropertyWrapper
to handle ContainIntrinsicSizeType.

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/animation/contain-intrinsic-size-interpolation-expected.txt:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::OptionalLengthPropertyWrapper::OptionalLengthPropertyWrapper):
(WebCore::OptionalLengthPropertyWrapper::canInterpolate const):
(WebCore::OptionalLengthPropertyWrapper::blend const):
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):

Canonical link: https://commits.webkit.org/256028@main
